flecs_ecs_tracing/src/metadata.rs

```diff
@@ -0,0 +1,157 @@
+//! Construct [`tracing_core::Metadata`] objects in the way required by that ecosystem.
+//!
+//! Dynamic call site stuff based on <https://github.com/slowli/tracing-toolbox/>
+
+use crate::util::*;
+use std::{
+    borrow::Cow,
+    collections::HashMap,
+    sync::{LazyLock, OnceLock, RwLock},
+};
+use tracing_core::{field::FieldSet, metadata::Level, Dispatch, Kind, Metadata};
+
+/// Registry of dynamic tracing metadata generated from Flecs
+pub(crate) static REGISTRY: LazyLock<RwLock<Registry>> = LazyLock::new(Default::default);
+
+/// Registry of dynamic tracing metadata generated from Flecs
+pub(crate) struct Registry {
+    pub(crate) metadata:
+        HashMap<MetadataRequest<'static>, &'static Metadata<'static>, fxhash::FxBuildHasher>,
+}
+
+impl Default for Registry {
+    fn default() -> Self {
+        Self {
+            // Capacity was chosen arbitrarily
+            metadata: HashMap::with_capacity_and_hasher(256, Default::default()),
+        }
+    }
+}
+
+pub(crate) fn init() {
+    LazyLock::force(&REGISTRY);
+}
+
+#[derive(Default)]
+pub(crate) struct DynamicCallsite {
+    /// Due to the opaque [`tracing_core::callsite::Identifier`],
+    /// which borrows [`tracing_core::Callsite`] for `'static`, have to use interior mutability.
+    ///
+    /// See [`MetadataRequest::register`] for details
+    pub(crate) metadata: OnceLock<Metadata<'static>>,
+}
+
+impl tracing_core::Callsite for DynamicCallsite {
+    fn set_interest(&self, _interest: tracing_core::Interest) {
+        // No-op
+    }
+
+    fn metadata(&self) -> &Metadata<'_> {
+        self.metadata.get().unwrap()
+    }
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
+pub(crate) enum MetadataKind {
+    // hash/eq skip filename/line since they may differ between start & stop
+    // only used when perf_trace is enabled
+    #[cfg_attr(not(feature = "perf_trace"), allow(unused))]
+    Span,
+    Event,
+}
+
+#[derive(Clone, Debug)]
+pub(crate) struct MetadataRequest<'a> {
```

Comment:
Going by code below, `filename` and `line` are only valid when `kind == MetadataKind::Event`, so it seems like `MetadataKind::Event` should be holding that data.

In that case, then the `PartialEq` and `Hash` would be derived and not have a manual implementation.

If there is a reason for not doing the expected thing (what I just described), then that should be documented with a comment somewhere.

Reply:
`filename` and `line` are valid for both `Event` and `Span`, but for `Span` they are not part of the hash key, since the push and pop events will come from different line numbers (& possibly different file names).

It's a moot point at the moment, though, since metadata for spans is currently not cached in the metadata registry, only in the span registry which uses the span name as the key. This may change in the future; an alternative design would have one metadata per `ecs_os_perf_trace_push` callsite and put the span names into values (like log messages).